# Assignment 10

## Problem 5.15
Another approach, known as Steffensen’s method, iterates according to

$$
x^{(k)}=x^{(k-1)}-\frac{f(x^{(k-1)})}{g(x^{(k-1)})}\\
g(x^{(k)})=\frac{f(x^{(k-1)}+f(x^{(k-1)}))-f(x^{(k-1)})}{f(x^{(k-1)})}
$$

This achieves quadratic convergence, at the cost of requiring two function evaluations per iteration. Implement it in Python for our example function $f(x) = e^{x− \sqrt{x}} − x$
and compare the required number of iterations to those of other methods.

In [7]:
from math import exp, sqrt, log

def f(x):
  return exp(x - sqrt(x)) - x

def g(x):
  return (f(x+f(x))-f(x))/f(x)

def steffensen(f, xold, kmax=200, tol=1.e-8):
    for k in range(1,kmax):
        xnew = xold - f(xold)/g(xold)
        xdiff = xnew - xold
        print("{0:2d} {1:1.16f} {2:1.16f}".format(k,xnew,xdiff))
        if abs(xdiff/xnew) < tol:
            break
        xold = xnew
    else:
        xnew = None
    return xnew

root = steffensen(f,0.8)
print('Root found:', root)
print()
root = steffensen(f,2.4)
print('Root found:', root)

 1 0.9922301653521665 0.1922301653521664
 2 0.9999850672219974 0.0077549018698310
 3 0.9999999999442540 0.0000149327222566
 4 1.0000000000000000 0.0000000000557460
Root found: 1.0

 1 2.5062337388360936 0.1062337388360937
 2 2.4912772884269212 -0.0149564504091724
 3 2.4909095337102896 -0.0003677547166316
 4 2.4909093169460590 -0.0000002167642306
 5 2.4909093169459848 -0.0000000000000742
Root found: 2.490909316945985


## Problem 5.16

Explore the roots of:
$$f(x) = −x^5 + 4x^4 − 4x^3 + x^2e^x − 2x^2 − 4xe^x + 8x + 4e^x − 8$$
using a method of your choice.

In [24]:
import numpy as np

def f(x):
  return -x**5 + 4*x**4 - 4*x**3 + x**2*np.exp(x) - 2*x**2 - 4*x*np.exp(x) + 8*x + 4*np.exp(x) - 8

def secant(f,x0,x1,kmax=200,tol=1.e-8):
  f0 = f(x0)
  for k in range(1,kmax):
    f1 = f(x1)
    ratio = (x1 - x0)/(f1 - f0)
    x2 = x1 - f1*ratio
    xdiff = abs(x2-x1)
    x0, x1 = x1, x2
    f0 = f1
    rowf = "{0:2d} {1:1.16f} {2:1.16f} {3:1.16f}"
    print(rowf.format(k,x2,xdiff,abs(f(x2))))
    if abs(xdiff/x2) < tol:
      break
  else:
    x2 = None
  return x2

In [34]:
root = secant(f,-5,-1)
print('Root found:', root)

 1 -1.0037719654448025 0.0037719654448025 5.6134116162744068
 2 -1.2835742599831090 0.2798022945383065 4.2244901576604548
 3 -1.1634244489228269 0.1201498110602821 1.1290668140537443
 4 -1.1887640810776250 0.0253396321547981 0.1574915758912070
 5 -1.1928716153619590 0.0041075342843340 0.0075118345294172
 6 -1.1926846185126867 0.0001869968492723 0.0000463268944504
 7 -1.1926857646886961 0.0000011461760094 0.0000000134919897
 8 -1.1926857650225993 0.0000000003339031 0.0000000000000178
Root found: -1.1926857650225993


In [33]:
root = secant(f,-2,0)
print('Root found:', root)

 1 -0.0783044237806349 0.0783044237806349 4.6425963408235003
 2 0.4874252703044413 0.5657296940850761 1.1157665660322085
 3 0.6664025157001625 0.1789772453957212 0.6202005492885805
 4 0.8903924225020410 0.2239899068018785 0.3322058761897324
 5 1.1487679071806574 0.2583754846786164 0.2620775519382761
 6 2.1143465835316917 0.9655786763510343 0.0414211837764284
 7 2.2956031325128694 0.1812565489811777 0.3641115205741841
 8 2.0910801265917800 0.2045230059210894 0.0253002040097989
 9 2.0758076916950619 0.0152724348967181 0.0170888568857812
10 2.0440238138117302 0.0317838778833317 0.0054622964486768
11 2.0290913697560282 0.0149324440557019 0.0023248310317214
12 2.0180265755012350 0.0110647942547932 0.0008756989133119
13 2.0113402091017547 0.0066863663994803 0.0003425307558338
14 2.0070445922227780 0.0042956168789767 0.0001311890774964
15 2.0043781138448997 0.0026664783778783 0.0000504340548133
16 2.0027128141248300 0.0016652997200697 0.0000193070577268
17 2.0016798832019265 0.001032930922903

In [32]:
root = secant(f,0,3)
print('Root found:', root)

 1 -2.4417723385936139 5.4417723385936139 249.4867743356549568
 2 2.8122668487559235 5.2540391873495373 5.0104613489315142
 3 2.7088269815087496 0.1034398672471739 3.4492318911710598
 4 2.4802967759364969 0.2285302055722527 1.2257815514356878
 5 2.3543087541083985 0.1259880218280984 0.5672340737097556
 6 2.2457900803949031 0.1085186737134953 0.2343368336771867
 7 2.1694003740988093 0.0763897062960939 0.0991983219800190
 8 2.1133265681499918 0.0560738059488175 0.0406183571910574
 9 2.0744459391278265 0.0388806290221653 0.0164430482300482
10 2.0480009386003384 0.0264450005274881 0.0065379899973905
11 2.0305454986757501 0.0174554399245883 0.0025694939640069
12 2.0192435729215061 0.0113019257542439 0.0010000459209820
13 2.0120420317172560 0.0072015412042501 0.0003867145021772
14 2.0075013538124176 0.0045406779048385 0.0001488722100618
15 2.0046592152303360 0.0028421385820816 0.0000571465717165
16 2.0028885164460126 0.0017706987843233 0.0000218957677234
17 2.0017886604172777 0.0010998560287

In [31]:
root = secant(f,4,5)
print('Root found:', root)

 1 4.1913657429610600 0.8086342570389400 45.7109930601608880
 2 4.3463947412245778 0.1550289982635178 38.0365789407339321
 3 5.1147625460875119 0.7683678048629341 297.4069214305254150
 4 4.4335214335970363 0.6812411124904756 29.1262935815634592
 5 4.4942871619517888 0.0607657283547525 20.3693449984140784
 6 4.6356329827231511 0.1413458207713623 10.2642704949030872
 7 4.5882728616810278 0.0473601210421233 1.8020330442276418
 8 4.5953458234061957 0.0070729617251679 0.1244632282534326
 9 4.5958705846657537 0.0005247612595580 0.0016890911226710
10 4.5958635584803504 0.0000070261854033 0.0000015500797304
11 4.5958635649223725 0.0000000064420220 0.0000000000196110
Root found: 4.5958635649223725


**Solution:** Using the secant method, we can find that the roots of $f(x)$ are $x=-1.192685$, $x=2$ and $x=4.59586$. For $x=-1.192685$ and $x=4.59586$, the method converges within around 10 iterations, meaning that the roots are of order 1. For $x=2$, the method takes around 40 iterations to converge meaning that the root is of a higher order.

This can further be verified by factorising the function,

$$
f(x) = (x-2)^2(-x^3+e^x-2)
$$

## Problem 5.17

Code up our companion matrix from Eq. (5.70) for a general polynomial. Then, use
it to find all the roots of our polynomial:
$$
f(x) = x^5 − 3x^4 + x^3 + 5x^2 − 6x + 2
$$

**Solution:** The companion matrix for $f(x)$ would be

$$
\begin{pmatrix}
0 & 1 & 0 & 0 & 0\\
0 & 0 & 1 & 0 & 0\\
0 & 0 & 0 & 1 & 0\\
0 & 0 & 0 & 0 & 1\\
-2 & 6 & -5 & -1 & 3
\end{pmatrix}
$$

In [17]:
import numpy as np

def mag(xs):
 return np.sqrt(np.sum(xs*xs))

def qrdec(A):
  n = A.shape[0]
  Ap = np.copy(A)
  Q = np.zeros((n,n))
  R = np.zeros((n,n))
  for j in range(n):
    for i in range(j):
      R[i,j] = Q[:,i]@A[:,j]
      Ap[:,j] -= R[i,j]*Q[:,i]
    R[j,j] = mag(Ap[:,j])
    Q[:,j] = Ap[:,j]/R[j,j]
  return Q, R

def qrmet(inA,kmax=100):
  A = np.copy(inA)
  for k in range(1,kmax):
    Q, R = qrdec(A)
    A = R@Q

  qreigvals = np.diag(A)
  return qreigvals

def companion(coeffs):
  n = len(coeffs)
  C = np.zeros((n,n))
  for i in range(0, n-1):
    C[i][i+1] = 1
  C[-1] = -np.array(coeffs)
  return C

# enter coefficients of the form (c0, c1, c2, c3, ...)
coeffs = (2., -6.,5.,1.,-3.)
C = companion(coeffs)
print('Companion Matrix:\n', C)
print()
qreigvals = qrmet(C, 100000)
print("Roots of the polynomial:")
print(qreigvals)

Companion Matrix:
 [[ 0.  1.  0.  0.  0.]
 [ 0.  0.  1.  0.  0.]
 [ 0.  0.  0.  1.  0.]
 [ 0.  0.  0.  0.  1.]
 [-2.  6. -5. -1.  3.]]

Roots of the polynomial:
[ 1.40909091 -1.40909091  1.00002     1.          0.99998   ]
